ColorPicker Updates & Fixes

robloo opened the pull request:
## What does the pull request do?

A roll-up of several fixes and a few improvements to the ColorPicker set of controls. See below for all details.

## What is the updated/expected behavior with this PR?

A roll-up of various ColorPicker updates and fixes:

 * Use SetCurrentValue() internally in ColorPicker controls
 * Fix flyout placement (after #10492 flyouts in the ColorPicker were opening downward instead of upward)
 * Match RGB/HSV color model control height with the hex input TextBox
 * Make ColorSpectrum.ThirdComponent a read-only DirectProperty (#10418)
 * Switch the color display name algorithm to work with the HSV color space. This will be much more accurate than the current algorithm which measures closest Euclidian distance in RGB color space (which doesn't match user perception of colors or the spectrums). 
    | old | new |
    |-|-|
    | ![image](https://user-images.githubusercontent.com/17993847/226135941-0f2369ea-126b-43f9-bee4-91263bf2d72c.png) | ![image](https://user-images.githubusercontent.com/17993847/226136003-f7ad0e7c-6559-4201-9f1f-272cb096a40a.png) |
 * Allow changing the Hex color format in the input box (#10094)
 * Support ColorSlider thumbs taller than the track (https://github.com/AvaloniaUI/Avalonia/pull/10439#issuecomment-1441602097)
    * Background spectrum is now calculated to follow the track size (instead of overall control size)
    * Lightweight styling resources were added to control control/track sizes independently without re-templating

## How was the solution implemented (if it's not obvious)?

See code and comments above

## Checklist

- ~[ ] Added unit tests (if possible)?~
- [x] Added XML documentation to any related classes?
- ~[ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation~

## Breaking changes

Yes, between previews. None since 0.10.
 * Make ColorSpectrum.ThirdComponent a read-only DirectProperty
 * The Hex input box default format was changed from ARGB (XAML/WinUI) to RGBA (CSS) by default. This can be switched back by setting `HexInputAlphaPosition` to `AlphaComponentPosition.Leading`.

## Obsoletions / Deprecations

None

## Fixed issues

Fixes #10094
Fixes #10418

robloo committed:
Use SetCurrentValue() in ColorPicker controls

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0030442-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

danipen commented:
Does it fix #10094 ?

rabbitism commented:
Is it possible to change the background size calculation in color slider? Currently it's impossible to restyle this slider to make thumb larger than tracking width.

robloo commented:
> Does it fix https://github.com/AvaloniaUI/Avalonia/issues/10094 ?

That is planned to be addressed; however, I still don't consider it a bug and its by design matching WinUI. Please comment on the original issue if you have any suggestions for the new enum.

> Is it possible to change the background size calculation in color slider? Currently it's impossible to restyle this slider to make thumb larger than tracking width.

I'll take a look at that when I can. Re-styling and re-templating is a major goal to make these prominitives usable for everyone. I know the WinUI design does this too so it definitely should be possible.

robloo committed:
Merge branch 'master' into colorpicker-updates-5

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0031219-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker-updates-5

robloo committed:
Restore ColorPicker flyout top placement

Default behavior was changed after https://github.com/AvaloniaUI/Avalonia/pull/10492

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0031677-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo committed:
Match color model control height with the hex TextBox

robloo committed:
Merge branch 'master' into colorpicker-updates-5

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0031998-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo committed:
Switch ColorHelper.ToDisplayName() to a hue-only algorithm

This is not working quite as accurately as expected so will likely be replaced by a hybrid technique next.

robloo committed:
Improve HsvColor saturation remarks

robloo committed:
Switch ColorSpectrum.ThirdComponent to a read-only DirectProperty

robloo committed:
Switch ColorHelper.ToDisplayName() to an HSV color space algorithm that uses all components

robloo committed:
Support configurable AlphaComponentPosition in ColorView

The default now matches CSS and differs from XAML/WinUI. The CSS trailing alpha component is in wider use for end-users now and it also matches the default slider ordering in the UI.

robloo commented:
@rabbitism I changed how ColorSlider work a bit to allow for styling like this:

![image](https://user-images.githubusercontent.com/17993847/226186952-fde78338-5b31-41cb-a351-b85c2c04b4e2.png)

Before, as you know, it was calculating the size of the spectrum background using the overall control bounds. This means it was always the same height as the thumb.

Now I am:

 1. By default calculating the background to match the size of the PART_Track template part. I think this is a safe assumption as the spectrum is 99% of the time meant to go on the slider track itself. PART_Track is already a template part of the base Slider so I don't need to add a new one either.
 2. Those restyling the control must be careful to make the background spectrum element (Anything binding to `{{TemplateBinding Background}}` the same height as the track. I didn't want to make a new template part so as a side-effect control theme authors just need to match the sizes themselves.
 3. To simplify setting overall height for the ColorSlider, I'm going to add a set of lightweight styling resources so everything doesn't need to be re-templated/re-themed to make small height/size adjustments.

Based on the above and the screenshot, is this what you need?

robloo committed:
Allow derived controls access to Slider template parts and state

robloo committed:
Calculate the ColorSlider spectrum based on the track size

robloo committed:
Update ColorSlider ControlTheme

1. Support lightweight resources to change track and control size independently
2. Make the background spectrum follow the track size

robloo committed:
Add custom ColorSliderThumbTheme to the sample page

This theme makes the thumb larger than the track

robloo committed:
Fix ColorHelper CI build error

robloo commented:
@maxkatz6 This one has a few important changes/fixes needed before 11.0 release. It also fixes a few things to allow re-theming sliders better (also necessary for 11.0 theme authors).

If there isn't a lot of time to review that's fine -- it would still be great to merge. I can take responsibility to fix any issues as they are reported.

robloo commented on src/Avalonia.Controls.ColorPicker/ColorView/AlphaComponentPosition.cs:
```diff
@@ -0,0 +1,26 @@
+    /// <summary>
+    /// Defines the position of a color's alpha component relative to all other components.
+    /// </summary>
+    public enum AlphaComponentPosition
```
Naming here was debated:

 * Orignially `AlphaPosition` but that is considered a bit too generically worded
 * An alternative `ColorAlphaPosition` is possible but perhaps "component" is better for what it does
 * Using `AlphaComponentPosition` matches with the `RgbComponent` and `HsvComponent` terminology

robloo commented on src/Avalonia.Controls.ColorPicker/ColorView/ColorView.Properties.cs:
```diff
@@ -42,6 +42,14 @@ public partial class ColorView
+        public static readonly StyledProperty<AlphaComponentPosition> HexInputAlphaPositionProperty =
+            AvaloniaProperty.Register<ColorView, AlphaComponentPosition>(
+                nameof(HexInputAlphaPosition),
+                AlphaComponentPosition.Trailing); // Match CSS (and default slider order) instead of XAML/WinUI
```
Note this is a breaking change from past behavior and now differs from the WinUI control and XAML default.

However, it is considered better to use #RRGGBBAA hex format:
 1. It matches CSS which is much more widely used now days (more user friendly and interoperable with other apps)
 2. It matches the order of the sliders in the default control theme (also more user friendly)

maxkatz6 replied:
I am afraid it might be confusing for XAML developers. But let's see how it goes. At least it is configurable.

robloo replied:
Yes, it might be. The assumption is it will be more friendly for app users (rather than devs) due to the above reasons. Following the UI and more widely used conventions is always a good thing. I could make a poll for it -- although it would be dev biased.

It can change during the RC phase too if there is lots of concern.

robloo replied:
I could also default the property to match XAML internally and then switch it in the control themes. I might do that in the future.

robloo commented on src/Avalonia.Controls.ColorPicker/Converters/ColorToHexConverter.cs:
```diff
@@ -62,21 +59,159 @@ public class ColorToHexConverter : IValueConverter
+        /// It is copied because it needs to be extended to handle alpha position.
+        /// However, the alpha position enum is only available in the controls namespace with the ColorPicker control.
+        /// </remarks>
+        private static bool TryParseHexFormat(
```
Not ideal to copy this method from Color. However, see remarks above for reasons why.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0032049-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

rabbitism commented:
@robloo the spectrum size change is what I need. Thanks!

maxkatz6 committed:
Merge branch 'master' into colorpicker-updates-5

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0032224-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]
